### PR TITLE
User can create an account

### PR DESCRIPTION
**Pull request by jonathanly**
#### Included in this PR:
- Adds user account creation feature spec
- Adds User controller
- Adds User#new, User#create and User#show controller actions.
- Adds views for new and show actions.
- Sets up and implements I18n in views

[Link](https://trello.com/c/zheQB8Fh/9-must-have-as-a-user-i-can-sign-up-for-an-account) to Trello card.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,3 +1,8 @@
+require "application_responder"
+
 class ApplicationController < ActionController::Base
+  self.responder = ApplicationResponder
+  respond_to :html
+
   protect_from_forgery with: :exception
 end
```
Comment (apalski) [inline app/controllers/application_controller.rb]: Elle suggested I move protect_from_forgery under private

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def create
+    @user = User.create(user_params)
+
+    respond_with @user
+  end
+  
+  def show
+    @user = User.find(params[:id])
```
Comment (apalski) [inline app/controllers/users_controller.rb]: When you get to edit and destroy it might be useful to move the find method into a #user method and add helper_method :user underneath the method

File: app/views/users/show.html.erb [hunk 3]
```diff
@@ -0,0 +1 @@
+<h1><%= "Welcome, #{@user.first_name}" %></h1>
```
Comment (apalski) [inline app/views/users/show.html.erb]: You will need to do translations here

File: app/views/users/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<h1>New Account</h1>
```
Comment (apalski) [inline app/views/users/new.html.erb]: You will need translation for this

File: config/locales/responders.en.yml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+    actions:
+      create:
+        notice: '%{resource_name} was successfully created.'
+        # alert: '%{resource_name} could not be created.'
+      update:
+        notice: '%{resource_name} was successfully updated.'
+        # alert: '%{resource_name} could not be updated.'
+      destroy:
+        notice: '%{resource_name} was successfully destroyed.'
+        alert: '%{resource_name} could not be destroyed.'
```
Comment (apalski) [inline config/locales/responders.en.yml]: Double quotes

File: config/locales/simple_form.en.yml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    # hints:
+    #   defaults:
+    #     username: 'User name to sign in.'
+    #     password: 'No special characters, please.'
+    # include_blanks:
+    #   defaults:
+    #     age: 'Rather not say'
+    # prompts:
+    #   defaults:
+    #     age: 'Select your age'
```
Comment (apalski) [inline config/locales/simple_form.en.yml]: double quotes

File: lib/templates/erb/scaffold/_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  <div class="form-inputs">
+  <%- attributes.each do |attribute| -%>
+    <%%= f.<%= attribute.reference? ? :association : :input %> :<%= attribute.name %> %>
+  <%- end -%>
+  </div>
+
+  <div class="form-actions">
+    <%%= f.button :submit %>
+  </div>
+<%% end %>
```
Comment (apalski) [inline lib/templates/erb/scaffold/_form.html.erb]: translations

File: spec/features/user_creates_new_account_spec.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+feature "user creates an account" do
+  scenario "successfully" do
+    visit(new_user_path)
+
+    fill_in("user_username", with: "user")
+    fill_in("user_email", with: "user@example.com")
+    fill_in("user_password", with: "password")
+    fill_in("user_first_name", with: "Foo")
+    fill_in("user_last_name", with: "Bar")
+    click_on("Create User")
```
Comment (apalski) [inline spec/features/user_creates_new_account_spec.rb]: translations

File: lib/templates/erb/scaffold/_form.html.erb (hunk 7 above)
Comment (jonathanly) [reply lib/templates/erb/scaffold/_form.html.erb]: Not too sure whether translations is applicable here.
It's the auto generated file from the simple_form library.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (jonathanly) [reply app/controllers/application_controller.rb]: Will do, thanks.
What was the reasoning behind this change?

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (jonathanly) [reply app/controllers/users_controller.rb]: Sounds like a good idea.
Will implement it when I add `edit` and `destroy` controller actions.

File: spec/features/user_creates_new_account_spec.rb (hunk 8 above)
Comment (jonathanly) [reply spec/features/user_creates_new_account_spec.rb]: Implemented with 37efc94

File: spec/features/user_creates_new_account_spec.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+    expect(page).to have_text("Welcome, Foo")
```
[the remark this answers is not in the document]
Comment (jonathanly) [reply spec/features/user_creates_new_account_spec.rb]: Implemented with 37efc94

File: config/locales/responders.en.yml (hunk 5 above)
Comment (jonathanly) [reply config/locales/responders.en.yml]: Fixed with 37efc94

File: app/views/users/new.html.erb (hunk 4 above)
Comment (jonathanly) [reply app/views/users/new.html.erb]: Implemented with 37efc94

File: app/views/users/show.html.erb (hunk 3 above)
Comment (jonathanly) [reply app/views/users/show.html.erb]: Implemented with 37efc94

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (jonathanly) [reply app/controllers/application_controller.rb]: Implemented with 37efc94

**Comment by jonathanly** [comment]
@elle Could I please get your review on this PR?

File: app/views/users/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+<h1><%= t(".title") %></h1>
+<%= simple_form_for @user do |f| %>
+  <%= f.input :username %>
+  <%= f.input :email %>
+  <%= f.input :password %>
+  <%= f.input :first_name %>
+  <%= f.input :last_name %>
```
Comment (elle) [inline app/views/users/new.html.erb]: What do you think about ordering the field in an order that might be friendlier to the user? For example:
first_name, last_name, and then the login fields...

File: app/views/users/show.html.erb [hunk 11]
```diff
@@ -0,0 +1 @@
+<h1><%= "#{t(".welcome")}, #{@user.first_name.capitalize}" %></h1>
```
Comment (elle) [inline app/views/users/show.html.erb]: A suggestion to:
1. Either handle capitalisation in CSS
2. Add a `to_s` method on user, which returns `first_name.capitalize`
This will allow you to just call `#{@user}` here

File: config/initializers/simple_form.rb [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+  # config.inputs_discovery = true
+
+  # Cache SimpleForm inputs discovery
+  # config.cache_discovery = !Rails.env.development?
+
+  # Default class for inputs
+  # config.input_class = nil
+
+  # Define the default class of the input wrapper of the boolean input.
+  config.boolean_label_class = 'checkbox'
```
Comment (elle) [inline config/initializers/simple_form.rb]: Double quotes

File: config/initializers/simple_form.rb [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+  config.default_wrapper = :default
+
+  # Define the way to render check boxes / radio buttons with labels.
+  # Defaults to :nested for bootstrap config.
+  #   inline: input + label
+  #   nested: label > input
+  config.boolean_style = :nested
+
+  # Default class for buttons
+  config.button_class = 'btn'
```
Comment (elle) [inline config/initializers/simple_form.rb]: Double quotes

File: config/initializers/simple_form.rb [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+  # Method used to tidy up errors. Specify any Rails Array method.
+  # :first lists the first message for each field.
+  # Use :to_sentence to list all errors for each field.
+  # config.error_method = :first
+
+  # Default tag used for error notification helper.
+  config.error_notification_tag = :div
+
+  # CSS class to add for error notification helper.
+  config.error_notification_class = 'error_notification'
```
Comment (elle) [inline config/initializers/simple_form.rb]: Double quotes

File: lib/templates/erb/scaffold/_form.html.erb (hunk 7 above)
Comment (elle) [reply lib/templates/erb/scaffold/_form.html.erb]: Yeah, auto-generated by Simple Form, so no reason to touch _in this case_

File: spec/controllers/users_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,4 @@
+require "rails_helper"
+
+RSpec.describe UsersController, type: :controller do
```
Comment (elle) [inline spec/controllers/users_controller_spec.rb]: Wil you add controller specs for the sad path? When for example invalid params are sent to the controller?

File: spec/features/user_creates_new_account_spec.rb [hunk 16]
```diff
@@ -0,0 +1,16 @@
+require "rails_helper"
+
+feature "user creates an account" do
```
Comment (elle) [inline spec/features/user_creates_new_account_spec.rb]: Side note: I usually try to match the feature description with the file name.

File: lib/templates/erb/scaffold/_form.html.erb (hunk 7 above)
Comment (apalski) [reply lib/templates/erb/scaffold/_form.html.erb]: Whoops, this is true!

File: config/initializers/simple_form.rb (hunk 14 above)
Comment (apalski) [reply config/initializers/simple_form.rb]: Might be time to implement Rubocop @jonathanly lol

File: spec/controllers/users_controller_spec.rb (hunk 15 above)
Comment (jonathanly) [reply spec/controllers/users_controller_spec.rb]: Implemented with 3b53af8

File: app/views/users/show.html.erb (hunk 11 above)
Comment (jonathanly) [reply app/views/users/show.html.erb]: Thank you @elle, I'll keep this in mind moving forwards when I know what I want to use throughout the app.

File: config/initializers/simple_form.rb (hunk 12 above)
Comment (jonathanly) [reply config/initializers/simple_form.rb]: Fixed with 8766dcb

File: config/initializers/simple_form.rb (hunk 13 above)
Comment (jonathanly) [reply config/initializers/simple_form.rb]: Fixed with 8766dcb

File: config/initializers/simple_form.rb (hunk 14 above)
Comment (jonathanly) [reply config/initializers/simple_form.rb]: Fixed with 8766dcb

File: spec/features/user_creates_new_account_spec.rb (hunk 16 above)
Comment (jonathanly) [reply spec/features/user_creates_new_account_spec.rb]: Changed with 4b5b99d

**Comment by jonathanly** [comment]
@elle, could you please look over my sad path in the feature spec.
I'm not too confident in my i18n implementation and whether my assertion is correct.

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def show
+    @user = User.find(params[:id])
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(:username, :email, :password, :first_name, :last_name)
```
Comment (elle) [inline app/controllers/users_controller.rb]: Line length is too long

A suggestion to implement Rubocop so that it will keep the style consistent.
Maybe in another PR??

File: config/locales/en.yml [hunk 18]
```diff
@@ -30,4 +30,19 @@
...
 en:
-  hello: "Hello world"
+  users:
+    new:
+      title: "Enter Your Details"
+      submit: "Sign Up"
+    show:
+      welcome: "Welcome"
+
+  activerecord:
```
Comment (elle) [inline config/locales/en.yml]: A question: why did you need to add the translations below?

File: spec/features/user_creates_new_account_spec.rb [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    fill_in("user_email", with: "user@example.com")
+    fill_in("user_password", with: "password")
+    fill_in("user_first_name", with: "Foo")
+    fill_in("user_last_name", with: "Bar")
+    click_on(I18n.t("users.new.submit"))
+
+    expect(page).to have_text("#{I18n.t("users.show.welcome")}, Foo")
+  end
+
+  scenario "unsuccessfully" do
```
Comment (elle) [inline spec/features/user_creates_new_account_spec.rb]: Because feature specs are expensive to run (the more you have, the slower your test suite will run), what we usually do is:
1. Do happy path specs in a feature spec to check that the app as a whole works as expected
2. Do sad path specs as controller specs (or other types of specs)

If you choose to leave the specs as is, then I would remove the controller spec file since it is empty.
And be aware that once the app grows, your test suite will take longer to run.

File: spec/features/user_creates_new_account_spec.rb [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+    fill_in("user_email", with: "user@example.com")
+    fill_in("user_password", with: "bad")
+    fill_in("user_first_name", with: "Foo")
+    fill_in("user_last_name", with: "Bar")
+    click_on(I18n.t("users.new.submit"))
+
+    expect(page).to have_text("#{I18n.t("activerecord.errors.models.user.too_short",
+      attribute: "Password",
+      count: 8)}"
+    )
```
Comment (elle) [inline spec/features/user_creates_new_account_spec.rb]: How about instead check that an alert flash is displayed on the page? and that you are back on the new page?

File: app/controllers/users_controller.rb (hunk 17 above)
Comment (jonathanly) [reply app/controllers/users_controller.rb]: Sure, I'll implement rubocop in the next PR.

File: config/locales/en.yml (hunk 18 above)
Comment (jonathanly) [reply config/locales/en.yml]: Should they be sorted alphabetically?

Comment (jonathanly) [reply config/locales/en.yml]: Oh, and translations we're added due to Annette's suggestions

File: spec/features/user_creates_new_account_spec.rb (hunk 20 above)
Comment (jonathanly) [reply spec/features/user_creates_new_account_spec.rb]: @elle, I've added an assertion to check that we're back on the new page with
```
expect(page).to have_current_path(new_user_path)
```
I can't seem to get this to go green. 
In the responders doc, it says that the new user template will be rendered if the form submission was a failure (due to model validations). However, the url path doesn't change and stays as `/users`

Comment (jonathanly) [reply spec/features/user_creates_new_account_spec.rb]: Update: Have decided to add sad path as controller specs instead.

File: config/locales/en.yml (hunk 18 above)
Comment (elle) [reply config/locales/en.yml]: These translations are the default AR ones. We don't need to add them unless you define other locales and need to translate these to the new langauge

File: spec/controllers/users_controller_spec.rb [hunk 21]
```diff
@@ -0,0 +1,31 @@
+require "rails_helper"
+
+describe UsersController do
+  context "POST #create" do
```
Comment (elle) [inline spec/controllers/users_controller_spec.rb]: 👍

File: spec/features/user_creates_new_account_spec.rb [hunk 22]
```diff
@@ -13,18 +13,4 @@
...
-
-    fill_in("user_username", with: "user")
-    fill_in("user_email", with: "user@example.com")
-    fill_in("user_password", with: "bad")
-    fill_in("user_first_name", with: "Foo")
-    fill_in("user_last_name", with: "Bar")
-    click_on(I18n.t("users.new.submit"))
-
-    expect(page).to have_text("#{I18n.t("flash.users.create.alert")}")
-    # expect(page).to have_current_path(new_user_path)
```
Comment (elle) [inline spec/features/user_creates_new_account_spec.rb]: Side note: I would write this assertion as: `expect(current_path).to eq new_user_path`

File: app/controllers/users_controller.rb [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+
+  private
+
+  def user_params
+    params.require(:user).permit(
+      :username,
+      :email,
+      :password,
+      :first_name,
+      :last_name
```
Comment (elle) [inline app/controllers/users_controller.rb]: Comma on last line

Comment (jonathanly) [reply app/controllers/users_controller.rb]: Fixed with ee4b34f

**Comment by jonathanly** [comment]
Thanks Annette and Elle for reviewing this branch! Merging it into master now.
